README.md
Amazon S3 Extended Über Edition
==========

[MODX Revolution 2.2.X] The Amazon S3 media source type extended.


CHANGELOG.md
Amazon S3 Extended Über Edition
==========
V0.0.4:
==========
Main:
- Added plugin that will cache all images on S3 on the first file upload.
  This will take some time, once the file has been uploaded it's done caching.
- Removed default image type. Image type will the same as the source file.
- Added debugMode property. (Default: "No")
- Added code function that will help the plugin.

Lexicon:
- Added English translation for debugMode description.
- Added English translation for code.

v0.0.3:
==========
Main:
- Added keep image ratio.
  You can enable this by setting the keepRatio property to "Yes".
- Added keepRatio property. (Default: "Yes")
- Added cacheToS3 property.  (Default: "Yes")
- Added s3CacheFolder property. (Default: "_cache")
- Added hideS3Cache property. (Default: "Yes")
- Added custom caching for the file browser.
- Images with spaces in their name should now be properly displayed in the file browser.
- Fixed downloading files with spaces in their names.

Lexicon:
- Added English translation for keepRatio description.
- Added English translation for cacheToS3 description.
- Added English translation for s3CacheFolder description.
- Added English translation for hideS3Cache description.


v0.0.2:
==========
Main:
- Added resize on upload for PNG, JPG, JPEG, GIF filetypes.
  You can enable this by setting the downSize property to "Yes".

- Added sanitize file on upload. (eg. "Foo Bar.PNG" => "foo-bar.png")
  You can enable this by setting the sanitizeFile property to "Yes".

- **downSize property now works as intended.**

- Added downSizeWidth property (in pixels). (Default: "300")
- Added downSizeHeight property (in pixels). (Default: "300")
- Added downSizeQuality property (in percents). (Default: "90")
- Added sanitizeFiles property. (Default: "Yes")

Lexicon:
- Added English translation for downSizeWidth description.
- Added English translation for downSizeHeight description.
- Added English translation for downSizeQuality description.
- Added Englihs translation for sanitizeFiles description.
- Added English translation for notImplemented error.


v0.0.1:
==========
Main:
- Added basic Amazon S3 source type
- ~~Added downSize property.~~ (Default: "No") Not working

Lexicon:
- Added English translation for package name.
- Added English translation for package description.
- Added English translation for url description.
- Added English translation for bucket description.
- Added English translation for key description.
- Added English translation for secret_key description.
- Added English translation for imageExtension description.
- Added English translation for thumbnailType description.
- Added English translation for thumbnailQuality description.
- Added English translation for skipFiles description.
- Added English translation for downSize description.

core/components/s3extended/model/s3extended/s3extended.functions.php
<?php
//////////////////////////////////////////////////////////////
///  S3 Extended by Robin Rijkeboer <robin@qaraqter.nl>     //
//   available at https://github.com/Qaraqter/s3extended   ///
//////////////////////////////////////////////////////////////
///                                                         //
// s3extended.functions.php - general functions class       //
//                                                         ///
//////////////////////////////////////////////////////////////
class s3extended_functions { 

    /** @var class $s3 */
    private $s3;

    public function __construct($s3) {
       $this->s3 = $s3;
    }

    public function generateFileManagerThumbs($url) {
        $s3CacheFolder = $this->s3->getOption('s3CacheFolder', $this->s3->properties, '_cache');
        $properties = $this->s3->getPropertyList();
        $s3CacheFolderURL = $properties['url']. $s3CacheFolder . "/";
        $cacheToS3 = $this->s3->getOption('cacheToS3', $this->s3->properties,'Yes');

        $imageWidth = $this->s3->ctx->getOption('filemanager_image_width', 400);
        $imageHeight = $this->s3->ctx->getOption('filemanager_image_height', 300);
        $thumbHeight = $this->s3->ctx->getOption('filemanager_thumb_height', 60);
        $thumbWidth = $this->s3->ctx->getOption('filemanager_thumb_width', 80);
        $thumbnailQuality = $this->s3->getOption('thumbnailQuality', $this->s3->properties, 90);

        $assetsPath = MODX_ASSETS_PATH . "components/s3extended";
        $pathinfo = pathinfo($url);
        $ext = strtolower($pathinfo['extension']);
        $tempFile = $assetsPath . "/tmp/" . uniqid();
        $cacheUrls = array();
        //Thumbnail
        $cacheFileName = md5($url) . $thumbWidth . "x" . $thumbHeight . "." . $ext;
        $cacheName = $assetsPath . "/cache/" . $cacheFileName;
        $cacheUrl = MODX_ASSETS_URL . "components/s3extended/cache/" . $cacheFileName;

        //SmallerImage
        $cacheFileName2 = md5($url) . $imageWidth . "x" . $imageHeight . "." . $ext;
        $cacheName2 = $assetsPath . "/cache/" . $cacheFileName2;
        $cacheUrl2 = MODX_ASSETS_URL . "components/s3extended/cache/" . $cacheFileName2;
        $url = str_replace(" ", "%20", $url);

        if ($cacheToS3 == "Yes"
            && $this->s3->driver->if_object_exists($this->s3->bucket, $s3CacheFolder . "/" . $cacheFileName)
            && $this->s3->driver->if_object_exists($this->s3->bucket, $s3CacheFolder . "/" . $cacheFileName2)) {
            array_push($cacheUrls, $s3CacheFolderURL . $cacheFileName);
            array_push($cacheUrls, $s3CacheFolderURL . $cacheFileName2);
            return $cacheUrls;
        }

        if (file_exists($cacheName) && file_exists($cacheName2)) {
            if ($cacheToS3 == "Yes") {
                if($this->uploadFile("", $s3CacheFolder, $cacheFileName, $cacheName)) {
                    $cacheUrl = $s3CacheFolderURL . $cacheFileName;
                    unlink($cacheName);
                }
                if($this->uploadFile("", $s3CacheFolder, $cacheFileName2, $cacheName2)) {
                    $cacheUrl2 = $s3CacheFolderURL . $cacheFileName2;
                    unlink($cacheName2);
                }
            }
            array_push($cacheUrls, $cacheUrl);
            array_push($cacheUrls, $cacheUrl2);
            return $cacheUrls;
        }

        if (!file_exists($assetsPath . "/tmp")) {
            $oldmask = umask(0);
            mkdir($assetsPath . "/tmp", 0777, true);
            umask($oldmask);
        }
        if (!file_exists($assetsPath . "/cache")) {
            $oldmask = umask(0);
            mkdir($assetsPath . "/cache", 0777, true);
            umask($oldmask);
        }
        
        if (file_put_contents($tempFile, fopen($url, 'r')) === false) {
            return 0;
        }
        
        list($width, $height) = getimagesize($tempFile);
        
        if ($width > $height) {
            $onePercent = $width / 100;
            $downSizePercentage = floatval(($imageWidth / $onePercent)) / 100;
            $imageHeight = $height * $downSizePercentage;
        } else {
            $onePercent = $width / 100;
            $downSizePercentage = floatval(($imageHeight / $onePercent)) / 100;
            $imageWidth = $width * $downSizePercentage;
        }
            // Load
            $thumb = imagecreatetruecolor($thumbWidth, $thumbHeight);
            $thumb2 = imagecreatetruecolor($imageWidth, $imageHeight);
            switch ($ext) {
                case "gif":
                    $source = imagecreatefromgif($tempFile);
                    imagecopyresampled($thumb, $source, 0, 0, 0, 0, $thumbWidth, $thumbHeight, $width, $height);
                    imagecopyresampled($thumb2, $source, 0, 0, 0, 0, $imageWidth, $imageHeight, $width, $height);
                    imagegif($thumb, $cacheName, $thumbnailQuality);
                    imagegif($thumb2, $cacheName2, $thumbnailQuality);
                    break;
                case "jpeg":
                    $source = imagecreatefromjpeg($tempFile);
                    imagecopyresampled($thumb, $source, 0, 0, 0, 0, $thumbWidth, $thumbHeight, $width, $height);
                    imagecopyresampled($thumb2, $source, 0, 0, 0, 0, $imageWidth, $imageHeight, $width, $height);
                    imagejpeg($thumb, $cacheName, $thumbnailQuality);
                    imagejpeg($thumb2, $cacheName2, $thumbnailQuality);
                    break;
                case "jpg":
                    $source = imagecreatefromjpeg($tempFile);
                    imagecopyresampled($thumb, $source, 0, 0, 0, 0, $thumbWidth, $thumbHeight, $width, $height);
                    imagecopyresampled($thumb2, $source, 0, 0, 0, 0, $imageWidth, $imageHeight, $width, $height);
                    imagejpeg($thumb, $cacheName, $thumbnailQuality);
                    imagejpeg($thumb2, $cacheName2, $thumbnailQuality);
                    break;
                case "png":
                    $source = imagecreatefrompng($tempFile);
                    imagecopyresampled($thumb, $source, 0, 0, 0, 0, $thumbWidth, $thumbHeight, $width, $height);
                    imagecopyresampled($thumb2, $source, 0, 0, 0, 0, $imageWidth, $imageHeight, $width, $height);
                    imagepng($thumb, $cacheName, $thumbnailQuality);
                    imagepng($thumb2, $cacheName2, $thumbnailQuality);
                    break;
                default:
                    $this->s3->xpdo->log(modX::LOG_LEVEL_ERROR, "[" . $this->s3->getTypeName() . "] " . $this->s3->xpdo->lexicon('s3extended.notImplemented') . " File: " . $pathinfo['basename']);
                    break;
            }
        unlink($tempFile);

        if ($cacheToS3 == "Yes") {
            if($this->uploadFile("", $s3CacheFolder, $cacheFileName, $cacheName)) {
                $cacheUrl = $s3CacheFolderURL . $cacheFileName;
                unlink($cacheName);
            }
            if($this->uploadFile("", $s3CacheFolder, $cacheFileName2, $cacheName2)) {
                $cacheUrl2 = $s3CacheFolderURL . $cacheFileName2;
                unlink($cacheName2);
            }
        }

        array_push($cacheUrls, $cacheUrl);
        array_push($cacheUrls, $cacheUrl2);
        return $cacheUrls;
    }

    public function uploadFile($parentContainer, $container, $fileName, $file) {
        $newPath = $parentContainer.rtrim($container,'/').'/';
        if (!$this->s3->driver->if_object_exists($this->s3->bucket, $newPath)) {
            $this->s3->createContainer($container, $parentContainer);
        }
        $newPath = $parentContainer.rtrim($container,'/').'/' . $fileName;
        $extension = pathinfo($fileName, PATHINFO_EXTENSION);
        $size = @filesize($file);
        $uploaded = $this->s3->driver->create_object(
            $this->s3->bucket,
            $newPath,
            array(
            'fileUpload' => $file,
            'acl' => AmazonS3::ACL_PUBLIC,
            'length' => $size,
            'contentType' => $this->s3->getContentType($extension),
        ));

        if (!$uploaded) {
            return false;
        }

        return true;
    }
}
